common/persistence/cassandra/queue_v2_store.go

```diff
@@ -233,27 +244,82 @@ func (q *queueV2Store) CreateQueue(
 	return &persistence.InternalCreateQueueResponse{}, nil
 }
 
-func (q *queueV2Store) getMinMessageID(
+func (s *queueV2Store) RangeDeleteMessages(
+	ctx context.Context,
+	request *persistence.InternalRangeDeleteMessagesRequest,
+) (*persistence.InternalRangeDeleteMessagesResponse, error) {
+	if request.InclusiveMaxMessageMetadata.ID < persistence.FirstQueueMessageID {
+		return nil, fmt.Errorf(
+			"%w: id is %d but must be >= %d",
+			persistence.ErrInvalidQueueRangeDeleteMaxMessageID,
+			request.InclusiveMaxMessageMetadata.ID,
+			persistence.FirstQueueMessageID,
+		)
+	}
+	q, err := s.getQueue(ctx, request.QueueType, request.QueueName)
+	if err != nil {
+		return nil, err
+	}
+	minMessageID := s.getPartitionMinMessageID(q.Metadata, request.QueueType, request.QueueName)
+	maxID := request.InclusiveMaxMessageMetadata.ID
+	err = s.session.Query(
+		TemplateRangeDeleteMessagesQuery,
+		request.QueueType,
+		request.QueueName,
+		0, // partition
+		minMessageID,
+		maxID,
+	).WithContext(ctx).Exec()
+	if err != nil {
+		return nil, gocql.ConvertError("QueueV2RangeDeleteMessages", err)
+	}
+	partitions := q.Metadata.Partitions
+	if len(partitions) != 1 {
+		partitionWarningMsg := s.getPartitionWarningMsg(
+			request.QueueType,
+			request.QueueName,
+			len(partitions),
+		)
+		s.logger.Warn(fmt.Sprintf(
+			"%s This query to delete messages assumes that there is only 1 partition. "+
+				"The messages are deleted, but we are not updating the min message ID of any partition.",
+			partitionWarningMsg,
+		))
+		return &persistence.InternalRangeDeleteMessagesResponse{}, nil
+	}
+	partitions[0].MinMessageId = maxID + 1
```

Comment:
hmmm, both min & max are inclusive seems a bit weird. Usually ranges are half open, either [inclusive, exclusive) or (exclusive, inclusive].

Reply:
I agree that ranges should be half-open normally, but I think the fact that this looks like a range is just a coincidence. It serves a different purpose. We need the min message ID to ensure that we skip as many tombstones as possible, and we need the max message ID to ensure that we insert messages with unique, monotonically increasing IDs. However, if you think of them as a range, both of them could be inclusive or exclusive. In the normal case, they're inclusive. However, they're exclusive sometimes due to race conditions and faults because we have to update the queue and messages out of sync.